### ISSUE TITLE
perf(GPU): Optimizer state (Adam m/v, BN running stats, weights) lives on host RAM — every train step round-trips through cuMemcpyHtoD/DtoH instead of staying GPU-resident

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#336]
## Problem

Even on the GPU training path (`TrainStepGpu` / `IAsyncGpuBackend.ExecuteDeferred`), the optimizer step happens on host CPU:

```
backend.ExecuteDeferred(scope => {
    var output = ForwardGpu(input);                           // forward stays on GPU
    var lossResult = LossFunction.CalculateLossAndGradientGpu(output, target);  // backward stays on GPU
    UpdateParametersGpu(config);                              // ← calls host-resident optimizer
});
```

`UpdateParametersGpu` walks `Layers`, fetches each layer's parameters (typically via `layer.GetParameters()` which returns CPU-resident `Tensor<T>` backed by managed `T[]`), runs the Adam moment update on CPU using `INumericOperations<T>`, and then writes the updated parameter back. On the next forward pass, the parameter is re-uploaded to a GPU buffer (or stays in the activation cache if the buffer is still alive; either way the round-trip is implicit).

The same pattern applies to Adam's `m` and `v` moment buffers (allocated via `Tensor<T>` → CPU), to BatchNorm running mean / variance, and to embedding tables for token-classification heads.

## Required behavior

Introduce a new `WeightLifetime.GpuPinned` value (parallel to the existing `Default`, `Streaming`, `GpuOffload`, `GpuManaged`). Tensors with this lifetime:

1. Allocate their backing storage directly on the active GPU backend via `IDirectGpuBackend.AllocateBuffer(byteCount)`, returning an `IGpuBuffer` reference rather than a managed `T[]`.
2. Are tracked by `WeightRegistry.RegisterWeight` like other lifetimes.
3. Expose `AsGpuBuffer<T>()` on `Tensor<T>` for callers that need to launch a kernel against the storage directly (cuBLAS, cuDNN).
4. Materialize lazily to host via `DeferredArrayMaterializer` (same pattern as `FinishGpuOp`) only when CPU code accesses the data.
5. Are pool-pinned — survive `TensorArena.Reset()` like the existing CPU `RentPinned` tier.

Then on the AiDotNet consumer side:

1. `LayerBase.EnsureInitialized` uses `TensorAllocator.RentPinnedOnGpu<T>` when `AiDotNetEngine.Current is DirectGpuTensorEngine`, falls back to CPU `RentPinned` otherwise.
2. `AdamOptimizer._tapeM` / `_tapeV` and `BatchNormalizationLayer._runningMean` / `_runningVariance` follow the same pattern.
3. The optimizer step kernel itself runs on GPU — adds a `GpuOptimizer` interface with cuBLAS-backed Adam/SGD/AdamW kernels, and `INumericOperations<T>` falls back to CPU only when the layer's parameters are CPU-resident.

## What this buys

For a model where the entire layer stack is `IGpuLayer`-compliant: the train loop never copies a parameter or moment between host and device. Wall-time savings dominate at small batch sizes where each per-op round-trip is amortized over fewer FLOPs.

## What is NOT in scope

- Multi-GPU parameter sharding (`ZeRO`, `FSDP`). Single-device, GPU-resident.
- Optimizer-state offload to host (the opposite direction — for memory-constrained large models). Tracked separately under #209.

## Tracking

Tensors-side API (the `WeightLifetime.GpuPinned` allocator, `RentPinnedOnGpu`, GPU optimizer kernel surface) lands in PR ooples/AiDotNet.Tensors#333.

Consumer wiring (LayerBase.EnsureInitialized, AdamOptimizer.\_tapeM/v, BatchNorm running stats) lands in a companion AiDotNet PR that bumps the Tensors NuGet to whatever version ships with #333.
